clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/NodeStateChangeChecker.java

```diff
@@ -196,86 +209,98 @@ private static Result checkIfStateSetWithDifferentDescription(NodeInfo nodeInfo,
-            // not increase if we allow node to go down
-            if (aGroupContainsNode(retiredAndNotUpGroups, node)) {
-                numberOfGroupsToConsider = retiredAndNotUpGroups.size() - 1;
-            }
-            if (numberOfGroupsToConsider < maxNumberOfGroupsAllowedToBeDown) {
-                log.log(FINE, "Allow, retiredAndNotUpGroups=" + retiredAndNotUpGroups);
-                return Optional.of(allow());
-            }
+        Set<Integer> retiredAndNotUpGroups = groupsWithNotRetiredAndNotUp(clusterState);
+        int numberOfGroupsToConsider = retiredAndNotUpGroups.size();
+        // Subtract one group if node is in a group with nodes already retired or not up, since number of such groups will
+        // not increase if we allow node to go down
+        if (aGroupContainsNode(retiredAndNotUpGroups, node)) {
+            numberOfGroupsToConsider = retiredAndNotUpGroups.size() - 1;
+        }
 
-            return Optional.of(disallow(String.format("At most %d groups can have wanted state: %s",
-                                                      maxNumberOfGroupsAllowedToBeDown,
-                                                      sortSetIntoList(retiredAndNotUpGroups))));
-        } else {
-            // Return a disallow-result if there is another node with a wanted state
-            var otherNodeHasWantedState = otherNodeHasWantedState(nodeInfo);
-            if (otherNodeHasWantedState.notAllowed())
-                return Optional.of(otherNodeHasWantedState);
+        var result = checkRedundancyForGroupsThatAreUp(retiredAndNotUpGroups);
+        if (result.isPresent() && result.get().notAllowed())
+            return result;
+
+        if (numberOfGroupsToConsider < maxNumberOfGroupsAllowedToBeDown) {
+            log.log(FINE, "Allow, retiredAndNotUpGroups=" + retiredAndNotUpGroups);
+            return Optional.of(allow());
+        }
+
+        return Optional.of(disallow(String.format("At most %d groups can have wanted state: %s",
+                                                  maxNumberOfGroupsAllowedToBeDown,
+                                                  sortSetIntoList(retiredAndNotUpGroups))));
+    }
+
+    private Optional<Result> checkRedundancyForGroupsThatAreUp(Set<Integer> retiredAndNotUpGroups) {
+        Set<Integer> groupsThatAreUp = groupsThatAreUp(retiredAndNotUpGroups);
+        log.log(FINE, "Check min replication for groups " + groupsThatAreUp);
+        for (int group : groupsThatAreUp) {
+            List<ConfiguredNode> nodesInGroup = getNodesInGroup(group);
+            for (var n : nodesInGroup) {
+                log.log(FINE, "Check min replication for index " + n.index() + " in group " + group);
+                Set<Integer> indexesToCheck = nodesInGroup.stream().map(ConfiguredNode::index).collect(Collectors.toSet());
+                var r = checkRedundancySeenFromDistributor(clusterInfo.getDistributorNodeInfo(n.index()), indexesToCheck);
```

Comment:
Are we guaranteed to check _all_ possible distributors that can have knowledge of buckets in the `indexesToCheck` set? Since we prune nodes from groups that are retired/not up. Bucket ownership and management happens _across_ groups and retired-status only applies to content nodes and not distributors, so I'm wondering if we'll risk skipping looking at the corresponding distributors on retired nodes. E.g. if we have 4 groups and 1 has all retired content nodes, 25% of the buckets will still be owned by (non-retired) distributors in that group.

Perhaps we should instead loop over all distributors that are in an available state rather than doing this by group?

Reply:
Yes, you might right about that, will look into it

Reply:
@vekterli Gave it a try, please have a look again